### PR TITLE
 feat(ais-dynamic-widgets): add implementation

### PR DESCRIPTION
**Pull request by Haroenv**
**Summary**

Adds a new widget _ais-experimental-dynamic-widgets_ that can be used to conditionally render other widgets. This condition is based on the search results, by default results.facetOrdering.facet.values, but can be overridden using transformItems

**Result**

```vue
<template>
  <ais-experimental-dynamic-widgets :transform-items="transformItems">
    <ais-refinement-list attribute="test1" />
    <ais-menu attribute="test2" />
    <ais-panel>
      <ais-hierarchical-menu :attributes="hierarchicalAttributes" />
    </ais-panel>
  </ais-experimental-dynamic-widgets>
</template>

<script>
export default {
  data() {
    return {
      hierarchicalAttributes: ["test3", "unused"],
    };
  },
  methods(_attributes, { results }) {
    // modified if you want to use userData instead of facet ordering
    return results.userData[0].ordering;
  },
};
</script>
```

See also:

- https://github.com/algolia/instantsearch.js/pull/4687
- DX-1665

### CODE REVIEW

File: src/components/DynamicWidgets.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  if (vnode.componentOptions && vnode.componentOptions.propsData) {
+    if (vnode.componentOptions.propsData.attribute) {
+      return vnode.componentOptions.propsData.attribute;
+    }
+    if (
+      vnode.componentOptions &&
+      Array.isArray(vnode.componentOptions.propsData.attributes)
+    ) {
+      return vnode.componentOptions.propsData.attributes[0];
+    }
```
Comment (shortcuts) [inline src/components/DynamicWidgets.js:15]:
```suggestion
    if (Array.isArray(vnode.componentOptions.propsData.attributes)) {
      return vnode.componentOptions.propsData.attributes[0];
    }
```

The first condition doesn't seem necessary

File: src/components/DynamicWidgets.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    // by default, render everything, but hidden so that the routing doesn't disappear
+    if (!this.state) {
+      const allComponents = [];
+      components.forEach(vnode => allComponents.push(vnode));
+
+      return createElement(
+        'div',
+        { attrs: { hidden: true }, class: [this.suit()] },
+        allComponents
+      );
```
Comment (shortcuts) [inline src/components/DynamicWidgets.js:71]: Is there a reason not to spread `components`, or directly pass a `components.map(...)`?

Comment (Haroenv) [reply src/components/DynamicWidgets.js:71]: components is a Map, so it can't be mapped directly

File: src/components/DynamicWidgets.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            [vnode]
+          )
+        );
+      }
+    });
+
+    // by default, render everything, but hidden so that the routing doesn't disappear
+    if (!this.state) {
+      const allComponents = [];
+      components.forEach(vnode => allComponents.push(vnode));
```
Comment (francoischalifour) [inline src/components/DynamicWidgets.js:62]: Any reasons to not use a single expression?

File: src/components/__tests__/DynamicWidgets.js [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+  expect(wrapper.html()).toMatchInlineSnapshot(`
+
+<div class="ais-DynamicWidgets">
+</div>
+
+`);
+});
+
+it('renders attributesToRender 1', () => {
```
Comment (francoischalifour) [inline src/components/__tests__/DynamicWidgets.js:97]: What does this test mean? Can its name reflect the actual scenario?

Comment (francoischalifour) [reply src/components/__tests__/DynamicWidgets.js:97]: A suggestion:

```suggestion
it('renders only the components with attributes that belong to `attributesToRender`', () => {
```

File: src/components/__tests__/DynamicWidgets.js [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+  });
+  expect(wrapper.html()).toMatchInlineSnapshot(`
+
+<div class="ais-DynamicWidgets">
+</div>
+
+`);
+});
+
+it('renders nothing with empty items', () => {
```
Comment (francoischalifour) [inline src/components/__tests__/DynamicWidgets.js:69]: Perhaps we can add a test with a panel header and empty items?

File: src/components/DynamicWidgets.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        { attrs: { hidden: true }, class: [this.suit()] },
+        allComponents
+      );
+    }
+
+    return createElement(
+      'div',
+      { class: [this.suit()] },
+      this.state.attributesToRender.map(attribute => components.get(attribute))
+    );
```
Comment (francoischalifour) [inline src/components/DynamicWidgets.js:75]: We don't seem to test the dynamic nature of this widget, represented by these lines.

File: src/components/DynamicWidgets.js (hunk 3 above)
Comment (Haroenv) [inline src/components/DynamicWidgets.js:62]:
```suggestion
      const allComponents = [...components.values()];
```

to check whether this works as an alternative that's shorter

Comment (Haroenv) [reply src/components/DynamicWidgets.js:62]: not possible with this babel config

File: src/components/DynamicWidgets.js (hunk 6 above)
Comment (Haroenv) [reply src/components/DynamicWidgets.js:74]: was really hard to get Vue to correctly recognise a change in data internally, but it works now: 2002f4b

File: src/components/DynamicWidgets.js (hunk 3 above)
Comment (Haroenv) [reply src/components/DynamicWidgets.js]: d9779b5

File: src/components/__tests__/DynamicWidgets.js (hunk 5 above)
Comment (Haroenv) [reply src/components/__tests__/DynamicWidgets.js:118]: 2002f4b

File: src/components/__tests__/DynamicWidgets.js (hunk 4 above)
Comment (Haroenv) [reply src/components/__tests__/DynamicWidgets.js]: 2002f4b

File: src/components/DynamicWidgets.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+            { key: attribute, class: [this.suit('widget')] },
+            [vnode]
+          )
+        );
+      }
+    });
+
+    // by default, render everything, but hidden so that the routing doesn't disappear
+    if (!this.state) {
+      const allComponents = Array.from(components.values());
```
Comment (francoischalifour) [inline src/components/DynamicWidgets.js:61]: Is using `Map.prototype.values()` OK given our browser support for VIS? ([see browser compatibility](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/values#browser_compatibility))

File: src/components/DynamicWidgets.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+function getVueAttribute(vnode) {
+  if (vnode.componentOptions && vnode.componentOptions.propsData) {
+    if (vnode.componentOptions.propsData.attribute) {
+      return vnode.componentOptions.propsData.attribute;
+    }
+    if (Array.isArray(vnode.componentOptions.propsData.attributes)) {
+      return vnode.componentOptions.propsData.attributes[0];
+    }
+  }
```
Comment (francoischalifour) [inline src/components/DynamicWidgets.js:13]: That's easier to read/understand. Wdyt?

```suggestion
  const props = vnode.componentOptions && vnode.componentOptions.propsData;

  if (props) {
    if (props.attribute) {
      return props.attribute;
    }
    if (Array.isArray(props.attributes)) {
      return props.attributes[0];
    }
  }
```

File: src/components/DynamicWidgets.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
+import { createWidgetMixin } from '../mixins/widget';
+import { EXPERIMENTAL_connectDynamicWidgets } from 'instantsearch.js/es/connectors';
+import { createSuitMixin } from '../mixins/suit';
+
+function getVueAttribute(vnode) {
```
Comment (francoischalifour) [inline src/components/DynamicWidgets.js:5]:
```suggestion
function getWidgetAttribute(vnode) {
```

File: src/components/DynamicWidgets.js [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+          createElement(
+            'div',
+            { key: attribute, class: [this.suit('widget')] },
+            [vnode]
+          )
+        );
+      }
+    });
+
+    // by default, render everything, but hidden so that the routing doesn't disappear
```
Comment (francoischalifour) [inline src/components/DynamicWidgets.js:59]: Won't this behavior result in an unexpected UI flash were some widget will appear and then get removed because computed based on `state.attributesToRender`?

What if we don't render anything when we don't have a state?

File: src/components/__tests__/DynamicWidgets.js [hunk 11]
```diff
@@ -0,0 +1,409 @@
...
+});
+
+it('renders attributesToRender (panel)', () => {
+  const localVue = createLocalVue();
+
+  localVue.use(plugin);
+
+  __setState({
+    attributesToRender: ['test2'],
+  });
```
Comment (francoischalifour) [inline src/components/__tests__/DynamicWidgets.js:236]: Can we created a similar test with this? (to ensure that the panel component is not added to the DOM)

```js
  __setState({
    attributesToRender: [],
  });
```

File: src/components/DynamicWidgets.js (hunk 10 above)
Comment (Haroenv) [reply src/components/DynamicWidgets.js:59]: note that the widgets we render initially are hidden, so this won't cause a flash, but possibly slightly delayed visuals compared to server side rendering (which solves this, as the widget information is available on that server + initial frontend render)

Comment (francoischalifour) [reply src/components/DynamicWidgets.js:59]: Right, I think I rather meant: won't this results in an unnecessary search parameters computation because it mounts widgets that will likely get removed?

Why don't we render nothing?

File: src/components/DynamicWidgets.js (hunk 7 above)
Comment (Haroenv) [reply src/components/DynamicWidgets.js:61]: right, only Map is IE11 compatible, what did you mean as alternative in https://github.com/algolia/vue-instantsearch/pull/922#discussion_r638579981 then?

File: src/components/DynamicWidgets.js (hunk 10 above)
Comment (Haroenv) [reply src/components/DynamicWidgets.js:59]: if we render nothing, the parameters of the URL can't be applied unfortunately, this was the best workaround I found

File: src/components/DynamicWidgets.js [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    vnode.componentOptions && vnode.componentOptions.children
+      ? vnode.componentOptions.children
+      : vnode.children;
+
+  if (Array.isArray(children)) {
+    // return first child with a truthy attribute
+    return children.reduce(
+      (acc, curr) => acc || getWidgetAttribute(curr),
+      undefined
+    );
```
Comment (eunjae-lee) [inline src/components/DynamicWidgets.js:26]:
```suggestion
    for (let child of children) {
      const attribute = getWidgetAttribute(child);
      if (attribute) {
        return attribute;
      }
    }
```

kinda old school but easier to read along with that comment, IMO.

File: src/components/__tests__/DynamicWidgets.js [hunk 13]
```diff
@@ -0,0 +1,420 @@
...
+  </div>
+</div>
+
+`);
+});
+
+it('renders nothing without children', () => {
+  __setState({
+    attributesToRender: [],
+  });
```
Comment (eunjae-lee) [inline src/components/__tests__/DynamicWidgets.js:103]: needed in this case?

File: src/components/__tests__/DynamicWidgets.js [hunk 14]
```diff
@@ -0,0 +1,420 @@
...
+  expect(wrapper.html()).toMatchInlineSnapshot(`
+
+<div class="ais-DynamicWidgets">
+</div>
+
+`);
+});
+
+it('renders nothing with empty items', () => {
```
Comment (eunjae-lee) [inline src/components/__tests__/DynamicWidgets.js:118]:
```suggestion
it('renders nothing with empty attributesToRender', () => {
```

File: src/components/__tests__/DynamicWidgets.js [hunk 15]
```diff
@@ -0,0 +1,420 @@
...
+    localVue,
+    propsData: {
+      transformItems: items => items,
+    },
+    slots: {
+      default: `<ais-refinement-list attribute="test1"/>`,
+    },
+    stubs: {
+      'ais-refinement-list': MockRefinementList,
+      'ais-menu': MockMenu,
```
Comment (eunjae-lee) [inline src/components/__tests__/DynamicWidgets.js:137]:
```suggestion
```

File: src/components/__tests__/DynamicWidgets.js [hunk 16]
```diff
@@ -0,0 +1,420 @@
...
+    >
+    </div>
+  </div>
+</div>
+
+`);
+
+  attributesToRender = ['test3'];
+
+  wrapper.vm.$forceUpdate();
```
Comment (eunjae-lee) [inline src/components/__tests__/DynamicWidgets.js:368]: can you try 

```js
await wrapper.setData({ attributesToRender: ['test3'] });
```

not sure if it works in this VTU version, though.

File: src/components/DynamicWidgets.js (hunk 12 above)
Comment (Haroenv) [reply src/components/DynamicWidgets.js:26]: for of needs transpilation, what's wrong with the reduce?

Comment (eunjae-lee) [reply src/components/DynamicWidgets.js:26]: didn't think of transpilation. nothing wrong with the reduce, just it took me several seconds to follow where it's going.

File: src/components/__tests__/DynamicWidgets.js (hunk 13 above)
Comment (Haroenv) [reply src/components/__tests__/DynamicWidgets.js:103]: Yes, that's what this test does, check if there's nothing in the return from attributesToRender, it doesn't render the initial widgets hidden anymore

File: src/components/__tests__/DynamicWidgets.js (hunk 16 above)
Comment (Haroenv) [reply src/components/__tests__/DynamicWidgets.js:368]: doesn't seem to work, i think because of the mock

Comment (Haroenv) [reply src/components/__tests__/DynamicWidgets.js:368]: doesn't seem to work, i think because of the mock

File: src/components/__tests__/DynamicWidgets.js [hunk 17]
```diff
@@ -0,0 +1,420 @@
...
+it('renders nothing without children', () => {
+  __setState({
+    attributesToRender: [],
+  });
+
+  const wrapper = mount(DynamicWidgets, {
+    propsData: {
+      transformItems: items => items,
+    },
+  });
```
Comment (eunjae-lee) [inline src/components/__tests__/DynamicWidgets.js:109]: If I understand correctly, this test is about not providing any children `slots: { default: ... }` and seeing if it renders an empty container. So setting attributesToRender to an empty array isn't relevant. With this, we don't know the reason why the empty container is there is because of attributesToRender: [] or no children.

Comment (Haroenv) [reply src/components/__tests__/DynamicWidgets.js]: solved in c0d1823

**Review by eunjae-lee — APPROVED** [review]
let's go!
